syntax = "proto3";

package categorizer;

service ExpenseCategorizer {
  rpc Predict (PredictRequest) returns (PredictResponse);
  rpc ForceRetrain (Empty) returns (StatusResponse);
  rpc GetStatus (Empty) returns (StatusResponse);
  rpc GetModelInfo (Empty) returns (ModelInfoResponse);
}

message PredictRequest {
  string text = 1;
  string user_id = 2;
}

message PredictionResult {
  string category_id = 1;
  string category_name = 2;
  string category_icon = 3;
  string category_color = 4;
  float confidence = 5;
}

message PredictResponse {
  PredictionResult primary = 1;
  repeated PredictionResult alternatives = 2;
  bool needs_confirmation = 3;
  string source = 4;
}

message StatusResponse {
  bool success = 1;
  string message = 2;
  int32 categories_count = 3;
  bool is_training = 4;
}

message ModelInfoResponse {
  string model_path = 1;
  string data_hash = 2;
  int32 categories_count = 3;
  bool is_training = 4;
  string metadata = 5;
}

message Empty {}